BOM inspection for cloud libraries with no BOM imported

etanshaul opened the pull request:
fixes #1969 

- Contains a quickfix that will add the latest BOM version. Will also strip out the version from the dependency definition if its defined.


Still having the same difficulty with setting up unit testing for these inspections. Will revisit this issue, but please review.

etanshaul commented:
Inspection warning:
![image](https://user-images.githubusercontent.com/1735744/38946171-adbd43fa-4306-11e8-9a0b-9b7164af199a.png)

Inspection details:
![image](https://user-images.githubusercontent.com/1735744/38946238-d6fd5a02-4306-11e8-8f73-8ca2c618da38.png)

Inspection entry:
![image](https://user-images.githubusercontent.com/1735744/38946269-f069c340-4306-11e8-826f-f44aa2ff7c93.png)

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudBomInspection.java:
```diff
@@ -0,0 +1,80 @@
+ * Base {@link DomElementsInspection} for inspecting google-cloud-java BOM related issues in the
+ * project's pom.xml.
+ */
+class CloudBomInspection extends DomElementsInspection<MavenDomProjectModel> {
```
Should we make it `abstract` to note this is not real implementation?

etanshaul replied:
yes good idea

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudBomInspection.java:
```diff
@@ -0,0 +1,80 @@
+
+  private static final Logger logger = Logger.getInstance(CloudBomInspection.class);
+
+  public CloudBomInspection() {
```
done

etanshaul commented:
Just found a bug where applying the quickfix to all instances (via the 'apply all..' link) will write the BOM multiple times. Working on a fix.

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudDependencyWithNoBomInspection.java:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    private void addBom(Project project) {
+      MavenProject mavenProject = MavenProjectsManager.getInstance(project).findProject(module);
```
Just to clarify a bit, `MavenProject` maps to `Module`, and there could be multiple `MavenProject`s per IDE `Project`, hence the need for the module field, correct?

etanshaul replied:
correct. this inspection in the scope of a particular pom.xml which is part of a particular module.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudDependencyWithNoBomInspection.java:
```diff
@@ -0,0 +1,170 @@
+    checkDependencyWithNoBom(domFileElement.getRootElement(), holder);
+  }
+
+  private void checkDependencyWithNoBom(
```
yes. I definitely should do that.

etanshaul commented:
@ivanporty I fixed the bug I mentioned and extracted common code for dependency visiting in the base class.

For the latter, I still had to leave some common logic in the implementation (for example, since one inspection checks for the presence of the BOM, the other check for a missing BOM). Its also arguable that I reduced readability with this refactoring. wdyt?

ivanporty commented:
I think it's pretty clean, maybe `Consumer` sticks out a bit. What if we provide couple of abstract methods to implement instead - one that decides if inspection needs to run on a POM, another that will take MavenDependency and matching CloudLibrary? then we can leave `checkFileElement` in base class only?

etanshaul commented:
but then aren't we left with some duplication like before? checking if there are no cloud libraries for example

also regarding

> another that will take MavenDependency and matching CloudLibrary

what did you mean exactly?

etanshaul commented:
I do like the idea of an abstract `shouldApplyInspection` (or similarly named) method though.

ivanporty commented:
I meant we keep common functionality and then we are left with two pieces:
* does this need to be run at all - i.e. `CloudDependencyWithNoBomInspection` checks for BOM presence
* if it does run, check for specific `MavenDomDependency` in an abstract method callback, something like 'abstract inspectAndFix(MavenDomDependency, Module)`

Then we won't need to override `checkFileElement` and provide consumer and it could be easier to read a single method (inspectAndFix)

etanshaul commented:
ah. got it. let me see how that looks.

etanshaul commented:
very nice suggestion. its a lot cleaner now. ptal
